src/cts/src/LatenciesBalancer.cpp

```diff
@@ -0,0 +1,512 @@
+                 3,
+                 "For node {}, isert {:2f} buffers",
+                 node->name,
+                 (worseDelay_ - node->delay) / root_->getTopBufferDelay());
+      node->nBuffInsert
+          = (int) ((worseDelay_ - node->delay) / root_->getTopBufferDelay());
+    }
+    return;
+  }
+
+  // If it is not a leaf node compute the amount of buffers needed for its
+  // children
+  std::map<int, std::vector<odb::dbITerm*>> buffersNeeded2Childern;
+  for (int child : node->childrenIds) {
+    balanceLatencies(child);
+    buffersNeeded2Childern[graph_[child].nBuffInsert].push_back(
+        graph_[child].inputTerm);
+  }
+
+  // If the children need a different amount of buffers insert this difference
+  std::vector<odb::dbITerm*> sinksInput;
+  int previouBufToInsert = 0;
+  int srcX, srcY;
+  if (node->inputTerm == nullptr) {
+    odb::dbNet* rootNet = root_->getTopInputNet();
+    odb::dbBTerm* clkInput = rootNet->get1stBTerm();
+    odb::Rect clkInputBBox = clkInput->getBBox();
+    srcX = clkInputBBox.xCenter();
+    srcY = clkInputBBox.yCenter();
+  } else {
+    node->inputTerm->getAvgXY(&srcX, &srcY);
+  }
+
+  for (auto it = buffersNeeded2Childern.rbegin();
+       it != buffersNeeded2Childern.rend();
+       ++it) {
+    int bufToInsert = it->first;
+    std::vector<odb::dbITerm*> children = it->second;
+    if (bufToInsert == -1) {
+      continue;
+    }
+
+    if (!previouBufToInsert) {
+      previouBufToInsert = bufToInsert;
+      sinksInput.clear();
+      sinksInput = children;
+      continue;
+    }
+
+    int numBuffers = previouBufToInsert - bufToInsert;
+    odb::dbITerm* delauBuffInput
+        = insertDelayBuffers(numBuffers, srcX, srcY, sinksInput);
+
+    sinksInput.clear();
+    sinksInput = children;
+    sinksInput.push_back(delauBuffInput);
+
+    previouBufToInsert = bufToInsert;
+  }
+
+  node->nBuffInsert = previouBufToInsert;
+}
+
+odb::dbITerm* LatenciesBalancer::insertDelayBuffers(
+    int numBuffers,
+    int srcX,
+    int srcY,
+    const std::vector<odb::dbITerm*>& sinksInput)
+{
+  // get bbox of current load pins without driver output pin
+  int maxSinkX = 0, maxSinkY = 0;
+  int minSinkX = std::numeric_limits<int>::max(),
+      minSinkY = std::numeric_limits<int>::max();
+  odb::dbNet* drivingNet = nullptr;
+  debugPrint(logger_,
+             CTS,
+             "insertion delay",
+             3,
+             "Inserting {} buffers for sinks:",
+             numBuffers);
+  for (odb::dbITerm* sinkInput : sinksInput) {
+    if (drivingNet == nullptr) {
+      drivingNet = sinkInput->getNet();
+    }
+    debugPrint(
+        logger_, CTS, "insertion delay", 3, "  {}", sinkInput->getName());
+    sinkInput->disconnect();
+    int sinkX, sinkY;
+    sinkInput->getAvgXY(&sinkX, &sinkY);
+    maxSinkX = std::max(maxSinkX, sinkX);
+    maxSinkY = std::max(maxSinkY, sinkY);
+    minSinkX = std::min(minSinkX, sinkX);
+    minSinkY = std::min(minSinkY, sinkY);
```

Comment:
It would be simpler to use a Rect and mergeInit() / merge() and xCenter/yCenter below.